lib/spack/spack/solver/asp.py

```diff
@@ -2026,33 +2032,74 @@ def _develop_specs_from_env(spec, env):
     spec.constrain(dev_info['spec'])
 
 
-#
-# These are handwritten parts for the Spack ASP model.
-#
-def solve(specs, dump=(), models=0, timers=False, stats=False, tests=False,
-          reuse=False):
-    """Solve for a stable model of specs.
-
-    Arguments:
-        specs (list): list of Specs to solve.
-        dump (tuple): what to dump
-        models (int): number of models to search (default: 0)
+class Solver(object):
+    """This is the main external interface class for solving.
+
+    It manages solver configuration and preferences in once place. It sets up the solve
+    and passes the setup method to the driver, as well.
+
+    Properties of interest:
+
+      ``reuse (bool)``
+        Whether to try to reuse existing installs/binaries
+
+      ``show (tuple)``
+        What information to print to the console while running. Options are:
+        * asp: asp program text
+        * opt: optimization criteria for best model
+        * output: raw clingo output
+        * solutions: models found by asp program
+        * all: all of the above
+
+      ``models (int)``
+        Number of models to search (default: 0 for unlimited)
+
+      ``timers (bool)``
+        Print out coarse fimers for different solve phases.
+
+      ``stats (bool)``
+        Print out detailed stats from clingo
+
+      ``tests (bool or tuple)``
+        If ``True``, concretize test dependencies for all packages. If
+        a tuple of package names, concretize test dependencies for named
+        packages. If ``False``, do not concretize test dependencies.
+
     """
-    driver = PyclingoDriver()
-    if "asp" in dump:
-        driver.out = sys.stdout
-
-    # Check upfront that the variants are admissible
-    for root in specs:
-        for s in root.traverse():
-            if s.virtual:
-                continue
-            spack.spec.Spec.ensure_valid_variants(s)
+    def __init__(self):
+        self.set_default_coanfiguration()
+
+    def set_default_coanfiguration(self):
+        # These properties are settable via spack configuration. `None`
+        # means go with the configuration setting; user can override.
+        self.reuse = None
+
+        # these are concretizer settings
+        self.dump = ()
+        self.models = 0
+        self.timers = False
+        self.stats = False
+        self.tests = False
+
+    def solve(self, specs):
+        driver = PyclingoDriver()
+        if "asp" in self.dump:
+            driver.out = sys.stdout
+
+        # Check upfront that the variants are admissible
+        for root in specs:
+            for s in root.traverse():
+                if s.virtual:
+                    continue
+                spack.spec.Spec.ensure_valid_variants(s)
 
-    setup = SpackSolverSetup()
-    return driver.solve(
-        setup, specs, dump, models, timers, stats, tests, reuse
-    )
+        if self.reuse is None:
```

Comment:
Why preserve a reuse option in the code in addition to the config? It doesn't look like the reuse options on any of these methods are actually hooked up to any code entrypoints, and I don't see them being used in testing either.

Reply:
This is a config property that users can set (which incidentally is how clingo manages this internally, b/c it has tons of options).  Generally when you have a *lot* of options, it's better to let the user modify a template (the defaults) than duplicate everything in every method signature.  I'm trying to avoid having these repeated everywhere in the code -- to reduce the places we have to modify to add new options.

Reply:
I think that the default should be pulled from config at `Solver` instantiation time, rather than in `Solver.solve`. If we use my `configuration_defaults` code from above, that dict could include `reuse: spack.config.get('concretizer:reuse', False')`